Sequence independent validator-transaction processor shutdown, restart

boydjohnson opened the pull request:
After this, python transaction processors can stop (if allowed to do so gracefully (1 keyboardinterrupt)) even during load. The tps can also recover after validator shutdown, even during load. The validator still needs at least 2 keyboard interrupts to stop it.

The protos are updated, so if seeking to test it, run protogen.

peterschwarz commented on protos/validator.proto:
```diff
@@ -33,59 +33,63 @@ message Message {
         TP_REGISTER_REQUEST = 1;
         // Registration response from the validator to the transaction processor
         TP_REGISTER_RESPONSE = 2;
+        // Tell the validator that the transaction processor won't take any more transactions
```
This reordering of the enums is certainly fine here, but we should probably come up with a strategy for giving ourselves room to grow.

boydjohnson replied:
I could add a break of 10 (20?) numbers between top level namespace message types. So TP would end at 12, but CLIENT wouldn't start until 23 (33?).

peterschwarz commented on sdk/python/sawtooth_sdk/client/future.py:
```diff
@@ -37,6 +44,9 @@ def result(self, timeout=None):
                 if not self._condition.wait(timeout):
                     raise FutureTimeoutError('Future timed out')
+        if isinstance(self._result, FutureError):
+            raise ValidatorConnectionError(
```
The error could be stored in the FutureError - keeping the error as part of the operation, not an interpretation of the error.

boydjohnson replied:
Yeah, sounds good. I will update this.

peterschwarz commented on sdk/python/sawtooth_sdk/client/stream.py:
```diff
@@ -46,17 +50,24 @@ class _SendReceiveThread(Thread):
+
+        :param url (str): the address to connect to the validator on
+        :param futures (FutureCollection):
+        :param ready_event (threading.Event:
```
Missing closing paren

boydjohnson replied:
Thanks, updating.

peterschwarz commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -33,65 +43,175 @@
+        fut = None
+        try:
+            self._register()
+            while True:
```
There probably could be a control-flag here, to be set during the shutdown phase.

boydjohnson replied:
So my thinking was that the transaction processor would only stop due to signals (sigint is handled by the KeyboardInterrupt exception handling) and not programmatically. As I see it, any exception bumps the flow of control out of the while loop, with sigint getting special handling.

peterschwarz commented on sdk/python/sawtooth_sdk/client/stream.py:
```diff
@@ -114,59 +124,124 @@ def _get_message(self):
+                asyncio.set_event_loop(self._event_loop)
+            if self._context is None:
+                self._context = zmq.asyncio.Context()
+            self._sock = self._context.socket(zmq.DEALER)
```
Should set the linger property

boydjohnson replied:
Thanks, updating.

vaporos commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -33,65 +44,174 @@
-            LOGGER.debug(
-                'received message of type: %s',
-                Message.MessageType.Name(msg.message_type))
+        :note: itertools.chain.from_iterable flattens the nested lists
```
Is this note useful as API doc, or should it be a non-API doc comment within the method?

vaporos commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -33,65 +44,174 @@
-            except InternalError as ie:
-                LOGGER.warning("State Error! %s", ie)
+            except ValidatorConnectionError as vce:
+                LOGGER.warning("%s", vce)
```
Can we provide more context here?  Tracking down where this was output would be challenging if we don't add any unique text.

vaporos replied:
Same comment for other logging statements here.

vaporos commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -33,65 +44,174 @@
+            except ValidatorConnectionError as vce:
+                LOGGER.warning("%s", vce)
+        except InternalError as ie:
+            LOGGER.warning("State Error! %s", ie)
```
This assumes all InternalErrors are state errors, but presumably there may be other reasons we have InternalErrors.

vaporos commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -33,65 +44,174 @@
+            except ValidatorConnectionError as vce:
+                LOGGER.warning("%s", vce)
+        except ValidatorConnectionError as vce:
+            LOGGER.warning("%s", vce)
```
It might be good to add comments on these except blocks that log errors to explain why we are only logging.

vaporos commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -33,65 +44,174 @@
+                        self._process_future(fut, 1, sigint=True)
+                        fut = self._stream.receive()
+            except TOError:
+                pass
```
pass statements that throw away exceptions always warrant a comment to explain.

vaporos commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -13,17 +13,28 @@
 # ------------------------------------------------------------------------------
 
+from concurrent.futures import CancelledError
+from concurrent.futures import TimeoutError as TOError
```
Renaming things on import generally makes it harder to read the code.  If this conflicts with another TimeoutError, we could instead refer to it fully qualified concurrent.futures.TimeoutError.

vaporos commented on protos/processor.proto:
```diff
@@ -50,6 +50,32 @@ message TpRegisterResponse {
 }
 
+// The unregistration request from the transaction processor to the validator/executor
+message TpUnRegisterRequest {
```
Another possible approach would be to send less information in this request, then find the processors based on identity instead of (family, version, encoding).  Presumably that approach would be more of an unregister-all message.

vaporos replied:
Seems like this should be TpUnregisterRequest since unregister is a word.

boydjohnson replied:
I like the approach of sending less and sending just one unregister message per request to disconnect even if there are multiple handlers. But now I can't think of anything that would need to be in the message since the zeromq identity is already known on the validator side.

vaporos commented on protos/processor.proto:
```diff
@@ -50,6 +50,32 @@ message TpRegisterResponse {
 
+// The unregistration request from the transaction processor to the validator/executor
+message TpUnRegisterRequest {
+    // the transaction family name for the logic of the handler
```
Capitalize the first t

vaporos commented on protos/processor.proto:
```diff
@@ -50,6 +50,32 @@ message TpRegisterResponse {
+    // the transaction family name for the logic of the handler
+    string family = 1;
+
+    // The version supported;
```
remove the trailing ;

vaporos commented on protos/processor.proto:
```diff
@@ -50,6 +50,32 @@ message TpRegisterResponse {
+    // The version supported;
+    string version = 2;
+
+    //supported encoding of the handler
```
Add space and capitalize the s

vaporos commented on protos/processor.proto:
```diff
@@ -50,6 +50,32 @@ message TpRegisterResponse {
+    repeated string namespaces = 4;
+}
+
+// a response sent from the validator to the transaction processor acknowledging the unregistration
```
Curious if the comments wrap < 80. Hard to judge from the github interface.

vaporos commented on protos/processor.proto:
```diff
@@ -50,6 +50,32 @@ message TpRegisterResponse {
+    //supported encoding of the handler
+    string encoding = 3;
+
+    // The namesapces that this transaction processor handler expects
```
Typo 'namesapces'.

Do we really need to specify the namespaces here?  I wouldn't think so, because they don't uniquely identify the TP in any way.

vaporos commented on sdk/python/sawtooth_sdk/client/stream.py:
```diff
@@ -26,14 +27,17 @@
 
 LOGGER = logging.getLogger(__file__)
 
+RECONNECT_EVENT = -1
```
A comment here describing why this is set to -1 would be helpful.

vaporos commented on sdk/python/sawtooth_sdk/client/stream.py:
```diff
@@ -46,17 +50,24 @@ class _SendReceiveThread(Thread):
+
+        :param url (str): the address to connect to the validator on
+        :param futures (FutureCollection):
+        :param ready_event (threading.Event):
```
Should fill out this API doc.  In particular, some description of ready_event works would be useful.

vaporos commented on sdk/python/sawtooth_sdk/processor/core.py:
```diff
@@ -33,65 +44,174 @@
+                    if fut is not None:
+                        # process futures as long as the tp has them,
+                        # if the message doesn't come from zeromq->asyncio
+                        # in 1 second raise a TimeOutError (TOError) and
```
What is "the message" in this context?  Why 1s?

peterschwarz commented:
You can add the break, but start the next block on an even number: TP ends
at 12, CLIENT starts at 100, etc.  Using http error codes as an example
here of the distances between blocks of related items.

On Mon, Feb 20, 2017 at 10:39 AM, Boyd Johnson <notifications@github.com>
wrote:

> *@boydjohnson* commented on this pull request.
> ------------------------------
>
> In protos/validator.proto
> <https://github.com/hyperledger/sawtooth-core/pull/99#discussion_r102055587>
> :
>
> > @@ -33,59 +33,63 @@ message Message {
>          TP_REGISTER_REQUEST = 1;
>          // Registration response from the validator to the transaction processor
>          TP_REGISTER_RESPONSE = 2;
> +        // Tell the validator that the transaction processor won't take any more transactions
>
> I could add a break of 10 (20?) numbers between top level namespace
> message types. So TP would end at 12, but CLIENT wouldn't start until 23
> (33?).
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperledger/sawtooth-core/pull/99#discussion_r102055587>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAROee_e-2qcR-eikOhjHbLWmk0paCP9ks5recGogaJpZM4ME3eQ>
> .
>

boydjohnson commented:
I updated logging statements, fixed typos, and added more comments. The largest difference is that one TpUnregisterRequest is sent for all handlers on a transaction processor, so the handling in processor_iterator.py and processor_handlers.py was updated.
